<a href="https://colab.research.google.com/github/aditisinghq/book-talk/blob/main/contentbased_NPL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [122]:
!pip install rake_nltk
#rake_nltk is the library i am  using to extract keywords from description!. it has stopwords (by default english ) preloaded

In [123]:
from rake_nltk import Rake
import nltk
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import CountVectorizer
df = pd.read_csv('/content/drive/MyDrive/books7k.csv', engine='python', error_bad_lines=False)
df.head()
df=df[['isbn10','title','authors','thumbnail','description','categories']]
df.dropna(inplace = True)
print(df.shape)
df.drop_duplicates(subset=['isbn10'])#there aren't any duplicate isbns(there are some anamolies)
print(df.shape)
df['description'].apply(type).value_counts()
import string
df['description'] = df['description'].str.replace('[{}]'.format(string.punctuation), '')
df['description'].astype(str)
df['description'].apply(type).value_counts()



(6247, 6)
(6247, 6)


<class 'str'>    6247
Name: description, dtype: int64

In [124]:
df['Key_words'] = ''
r = Rake()
for index, row in df.iterrows():
  r.extract_keywords_from_text(row['description'])   # to extract key words, default in lower case
  key_words_dict_scores = r.get_word_degrees()   
  row['Key_words'] = list(key_words_dict_scores.keys())   # to assign list of key words to new column


df.head()


isbn10  ...                                          Key_words
0  0002005883  ...  [tale, best, friend, ’, life, style, sons, pit...
1  0002261987  ...  [ones, superb, storytelling, opened, first, dr...
2  0006163831  ...  [volume, two, stephen, donaldsons, acclaimed, ...
3  0006178731  ...  [hate, mafias, schemes, win, organized, crime,...
4  0006280897  ...  [natural, loves, become, distorted, without, l...

[5 rows x 7 columns]

In [125]:
#adding the categories i.e. genres to keywords as well so that similar genre books can be selected
df['categories']= df['categories'].apply(lambda x:x.split())
df['categories'].astype(str)
df['Key_words']=df['categories']+df['Key_words']
df.head()


isbn10  ...                                          Key_words
0  0002005883  ...  [Fiction, tale, best, friend, ’, life, style, ...
1  0002261987  ...  [Detective, and, mystery, stories, ones, super...
2  0006163831  ...  [American, fiction, volume, two, stephen, dona...
3  0006178731  ...  [Fiction, hate, mafias, schemes, win, organize...
4  0006280897  ...  [Christian, life, natural, loves, become, dist...

[5 rows x 7 columns]

**making authors names into keywords as well, each author's name will be merged(firstnamelastname) and added as a separate keyword. **

In [126]:
df['authors']= df['authors'].apply(lambda x:x.lower().replace(' ',''))

In [127]:
df.head()

isbn10  ...                                          Key_words
0  0002005883  ...  [Fiction, tale, best, friend, ’, life, style, ...
1  0002261987  ...  [Detective, and, mystery, stories, ones, super...
2  0006163831  ...  [American, fiction, volume, two, stephen, dona...
3  0006178731  ...  [Fiction, hate, mafias, schemes, win, organize...
4  0006280897  ...  [Christian, life, natural, loves, become, dist...

[5 rows x 7 columns]

In [128]:
df['authors']= df['authors'].apply(lambda x:x.lower().replace(';',' '))
df.head()

isbn10  ...                                          Key_words
0  0002005883  ...  [Fiction, tale, best, friend, ’, life, style, ...
1  0002261987  ...  [Detective, and, mystery, stories, ones, super...
2  0006163831  ...  [American, fiction, volume, two, stephen, dona...
3  0006178731  ...  [Fiction, hate, mafias, schemes, win, organize...
4  0006280897  ...  [Christian, life, natural, loves, become, dist...

[5 rows x 7 columns]

In [129]:
df

isbn10  ...                                          Key_words
0     0002005883  ...  [Fiction, tale, best, friend, ’, life, style, ...
1     0002261987  ...  [Detective, and, mystery, stories, ones, super...
2     0006163831  ...  [American, fiction, volume, two, stephen, dona...
3     0006178731  ...  [Fiction, hate, mafias, schemes, win, organize...
4     0006280897  ...  [Christian, life, natural, loves, become, dist...
...          ...  ...                                                ...
6803  8173031010  ...  [Adventure, stories, tale, 1932, enlightenment...
6804  817992162X  ...  [Health, &, Fitness, wisdom, confront, peace, ...
6805  8185300534  ...  [Philosophy, helps, suffering, unique, teacher...
6808  9027712050  ...  [History, logical, thoroughly, reliable, ba, s...
6809  9042003405  ...  [Literary, Criticism, host, rewarding, collect...

[6247 rows x 7 columns]

In [130]:
df['authors']= df['authors'].apply(lambda x:x.split())
df['authors'].astype(str)
df['Key_words']=df['authors']+df['Key_words']
df.head()


isbn10  ...                                          Key_words
0  0002005883  ...  [marilynnerobinson, Fiction, tale, best, frien...
1  0002261987  ...  [charlesosborne, agathachristie, Detective, an...
2  0006163831  ...  [stephenr.donaldson, American, fiction, volume...
3  0006178731  ...  [sidneysheldon, Fiction, hate, mafias, schemes...
4  0006280897  ...  [clivestapleslewis, Christian, life, natural, ...

[5 rows x 7 columns]

In [131]:
#creating the bag of words which is used by the rake function to get vectors
df['Bag_of_words'] = ''
for index, row in df.iterrows():
    words = ''
    words += ' '.join(row['Key_words']) + ' '
    row['Bag_of_words'] = words
df['Bag_of_words'] = df['Bag_of_words'].str.strip().str.replace('   ', ' ').str.replace('  ', ' ')


In [132]:
df.head(7)

isbn10  ...                                       Bag_of_words
0  0002005883  ...  marilynnerobinson Fiction tale best friend ’ l...
1  0002261987  ...  charlesosborne agathachristie Detective and my...
2  0006163831  ...  stephenr.donaldson American fiction volume two...
3  0006178731  ...  sidneysheldon Fiction hate mafias schemes win ...
4  0006280897  ...  clivestapleslewis Christian life natural loves...
5  0006280935  ...  clivestapleslewis Christian life renowned chri...
6  0006353282  ...           agathachristie Authors, English donation

[7 rows x 8 columns]

In [133]:
#trial
df['Bag_of_words'][3]


'sidneysheldon Fiction hate mafias schemes win organized crime dangerous dramatic world implacable enemy love flashbulb lit courtrooms ambitious prosecutors begin way hatred destructive memorable mesmerizing heroine jennifer brilliant beautiful climb dark arena political power attorney'

In [134]:
df=df.reset_index()

In [135]:
count = CountVectorizer()
count_matrix = count.fit_transform(df['Bag_of_words'])
count_matrix


<6247x38388 sparse matrix of type '<class 'numpy.int64'>'
	with 237350 stored elements in Compressed Sparse Row format>

In [136]:
cosine_sim = cosine_similarity(count_matrix, count_matrix)
print(cosine_sim)

[[1.         0.01059998 0.02649995 ... 0.03623939 0.00979968 0.        ]
 [0.01059998 1.         0.025      ... 0.03418817 0.01849001 0.02581989]
 [0.02649995 0.025      1.         ... 0.         0.02311251 0.        ]
 ...
 [0.03623939 0.03418817 0.         ... 1.         0.03160698 0.02942449]
 [0.00979968 0.01849001 0.02311251 ... 0.03160698 1.         0.        ]
 [0.         0.02581989 0.         ... 0.02942449 0.         1.        ]]


In [137]:
indices = pd.Series(df['isbn10'])
indices[:5]


0    0002005883
1    0002261987
2    0006163831
3    0006178731
4    0006280897
Name: isbn10, dtype: object

In [138]:
df

index  ...                                       Bag_of_words
0         0  ...  marilynnerobinson Fiction tale best friend ’ l...
1         1  ...  charlesosborne agathachristie Detective and my...
2         2  ...  stephenr.donaldson American fiction volume two...
3         3  ...  sidneysheldon Fiction hate mafias schemes win ...
4         4  ...  clivestapleslewis Christian life natural loves...
...     ...  ...                                                ...
6242   6803  ...  hermannhesse Adventure stories tale 1932 enlig...
6243   6804  ...  robinsharma Health & Fitness wisdom confront p...
6244   6805  ...  srinisargadattamaharaj sudhakars.dikshit Philo...
6245   6808  ...  georgwilhelmfriedrichhegel History logical tho...
6246   6809  ...  helenagrice timwoods Literary Criticism host r...

[6247 rows x 9 columns]

**we are searching the book by isbn10 here but in the final code will use fuzzywuzzy to search through entering book title like done in the colab_filter **

In [139]:
def recommend(isbn, cosine_sim = cosine_sim):
    recommended_books = []
    resdf=pd.DataFrame()
    idx = indices[indices == isbn].index[0]   # to get the index of the movie title matching the input movie
    score_series = pd.Series(cosine_sim[idx]).sort_values(ascending = False)  
    top_10_indices = list(score_series.iloc[1:11].index)   # to get the indices of top 10 most similar movies
    # [1:11] to exclude 0 (index 0 is the input movie itself)
    for i in top_10_indices:   # to append the titles of top 10 similar movies to the recommended_movies list
        recommended_books.append(list(df['isbn10'])[i])
    resdf=df[df['isbn10'].isin(recommended_books)]  
    return resdf

recommend('0439064864')# try 0439064864 for harry potter 2, 0439064864 is the time machine by hg wells

#df.loc[df['isbn10'] == '0451528557']
#print(df['title'][2510])#so duplicate book titles exist!

index  ...                                       Bag_of_words
454     473  ...  maudhartlovelace Juvenile Fiction wonderful su...
2503   2676  ...  rowling,j.k. Juvenile Fiction ministry looming...
2519   2697  ...  j.k.rowling marygrandpre Juvenile Fiction dark...
2520   2698  ...  rowling,j.k. Juvenile Fiction closest friends ...
2531   2710  ...  rowling,j.k. Juvenile Fiction father ’ youknow...
2533   2712  ...  j.k.rowling Juvenile Fiction azkaban phoenix s...
2542   2723  ...  rowling,j.k. Juvenile Fiction named tom riddle...
2548   2730  ...  j.k.rowling Juvenile Fiction azkaban order hal...
4169   4508  ...  j.k.rowling Juvenile Fiction yet preparing har...
5433   5911  ...  marissamoss Juvenile Fiction older sister also...

[10 rows x 9 columns]

**conclusion**
we can see all of these are similar movies, share categories and from the description we can tell that they are science fiction books and have the author H.G. wells in common in one some too. 

**trying to find how many books are in common in Books and books7k**


In [140]:
bdf = pd.read_csv('/content/drive/MyDrive/Books.csv', engine='python', error_bad_lines=False)
bdf.head()

ISBN  ...                                        Image-URL-L
0  0195153448  ...  http://images.amazon.com/images/P/0195153448.0...
1  0002005018  ...  http://images.amazon.com/images/P/0002005018.0...
2  0060973129  ...  http://images.amazon.com/images/P/0060973129.0...
3  0374157065  ...  http://images.amazon.com/images/P/0374157065.0...
4  0393045218  ...  http://images.amazon.com/images/P/0393045218.0...

[5 rows x 8 columns]

In [141]:
in_common=df[df['isbn10'].isin(list(bdf['ISBN'].unique()))]
in_common.shape

(2265, 9)

this means that we can only use books out of 2265 books in collaborative filtering or to search. 